modules/bmc/redfish.rb

```diff
@@ -0,0 +1,235 @@
+require 'redfish_client'
+require 'bmc/base'
+require 'bmc/redfish/dell'
+require 'bmc/redfish/hpe'
+
+module Proxy
+  module BMC
+    class Redfish < Base
+      include Proxy::Log
+      include Proxy::Util
+
+      def initialize(args)
+        @redfish_verify_ssl = Proxy::BMC::Plugin.settings.bmc_redfish_verify_ssl
+        # enforce SSL verification as a default, if the value isn't set in the config
+        # and bail if it's set to something nonsensical
+        @redfish_verify_ssl = true if @redfish_verify_ssl.nil?
+        raise ::Proxy::Error::ConfigurationError.new("bmc_redfish_verify_ssl must be boolean (true/false)") unless [true, false].include?(@redfish_verify_ssl)
+        super
+        load_vendor_overrides
+      end
+
+      def connect(args = { })
+        connection = RedfishClient.new("https://#{args[:host]}/", verify: @redfish_verify_ssl)
+        connection.login(args[:username], args[:password])
+        connection
```

Comment:
You could memorize the connection, it's a Ruby pattern that is quite common.

Reply:
I had originally thought I would do just that, but as it happens (assuming I've understood the rest of the code correctly, which maybe I haven't) it wouldn't serve any purpose -- every call to the proxy instantiates a new BMC object, for which connect is only called once. It looks to me like an extremely invasive change.

The redfish gem itself does its own memoizing, and also provides some caching. It keeps its auth token with the object, and will not perform a subsequent remote call to the BMC to reference data it's already seen. I ran into some problems with the cache needing flushed manually after calls that would update BMC data (e.g. changing the one-time boot device), but given the way the Proxy instantiates BMC objects it wasn't necessary to account for cache flushing in the plugin (a new BMC object with a new connection and an empty cache will be created for the next BMC Proxy request).

I think the main incentive for doing a bigger change to reuse the BMC objects across multiple Proxy calls would be to reduce the impact of network latency on querying this data, as currently there are quite a few round trips needed to query all the data needed to, for example, render a Host page in the Foreman UI. Any non-trivial latency is immediately obvious as it can take 30-40 seconds for all that data to return for some of our most distant subnets. But the work that's being done to put subnet associations on the BMC Proxy ought to take care of maybe 90% of that, and my sense is that's probably good enough given the effort that would need to go into redoing the app and the proxy to support longer lived BMC connections.

On the other hand I could easily be wrong about any or all of this. I don't mind finding out.